BranchSDK/tools/build_wix.py

```diff
@@ -0,0 +1,304 @@
+def file_identifier(path):
+    build_root = os.path.abspath(os.path.dirname(__file__) + "../../build")
+    # Wix error message:
+    #  Identifiers may contain ASCII characters A-Z, a-z, digits,
+    #  underscores (_), or periods (.).  Every identifier must
+    #  begin with either a letter or an underscore.
+    return path.replace(build_root + os.sep, "").replace(os.sep, ".")
+
+# Find all subdirectories to all depths. Returns a flat list, for
+# a ComponentGroup.
+def all_subdirs(path):
+    if not os.path.isdir(path):
+        return []
+
+    # Now path must be a directory
+
+    # List of full paths to all immediate children of path
+    all_paths = [os.path.join(path, f) for f in os.listdir(path)]
+
+    # Filter out only directories.
+    # List of full paths to all immediate subdirectories of path
+    all_dirs = [f for f in all_paths if os.path.isdir(f)]
+
+    # Now all_nested is a list of lists. Flatten it out. There's no
+    # possibility of repetition of elements.
+    result = all_dirs
+    for d in all_dirs:
+        nested = all_subdirs(d)
+        result = result + nested
+    return result
+
+# -----
+# ----- Conveniences to make specific ET elements
+# -----
+
+def make_directory_elem(elem, identifier, name=None):
+    if name:
+        return SubElement(elem, "Directory", {"Id": identifier, "Name": name})
+    else:
+        return SubElement(elem, "Directory", {"Id": identifier})
+
+def make_component_elem(elem, identifier, directory):
+    # TODO: Generating a new UUID each time here. If that's an issue,
+    # can put these in a JSON file and add if/when new directories
+    # appear.
+    return SubElement(elem, "Component", {
+        "Id": identifier,
+        "Directory": directory,
+        "Guid": str(uuid4())
+        })
+
+def make_file_elem(elem, identifier, source):
+    return SubElement(elem, "File", {"Id": identifier, "Source": source})
+
+# -----
+# ----- Methods to generate Wix XML from the contents of a given path
+# -----
+
+"""
+Produces a nested tree of Directory elements like:
+
+          <Directory Id="INCLUDEFOLDER" Name="include">
+            <Directory Id="BRANCHIOINCLUDEFOLDER" Name="BranchIO">
+              <Directory Id="BRANCHIOEVENTINCLUDEFOLDER" Name="Event" />
+            </Directory>
+          </Directory>
+"""
+def wix_directory(elem, path):
+    basepath = os.path.basename(path)
+    directory_id = file_identifier(path)
+    wdir = make_directory_elem(elem, directory_id, basepath)
+
+    files = os.listdir(path)
+    for f in files:
+        full_path = os.path.join(path, f)
+        if not os.path.isdir(full_path):
+            continue
+        wix_directory(wdir, full_path)
+
+    return wdir
+
+"""
+Generates a single Component element with all child File elements.
+Used by wix_components.
+"""
+def wix_component(elem, path, identifier=None):
+    # TODO: Determine a better Id (2nd arg to make_component_elem)
```

Comment:
These were previously more descriptive like BranchHeaders, BranchEventHeaders, but it's harder to generate that automatically, so went with this for now.